File: app/android/app/src/main/kotlin/dev/kishorek/relay/wifi/LanDiscoverer.kt
package dev.kishorek.relay.wifi

import android.content.Context
import android.net.ConnectivityManager
import android.net.Network
import android.net.NetworkCapabilities
import android.net.NetworkRequest
import android.net.nsd.NsdManager
import android.net.nsd.NsdServiceInfo
import java.util.ArrayDeque

/**
 * mDNS discovery over Android's NsdManager.
 *
 * Plain Android, no Flutter: everything Pigeon-shaped lives in
 * [MeshDiscoveryPlugin]. Two things here are less obvious than they look.
 *
 * First, resolves are serialised. `NsdManager.resolveService` fails with
 * `FAILURE_ALREADY_ACTIVE` if a second resolve starts before the first
 * finishes, and a crowded network produces found-callbacks far faster than
 * resolves complete. Without the queue most peers are simply never resolved,
 * on exactly the busy networks where the transport matters most.
 *
 * Second, an unresolved service is useless. `onServiceFound` carries a name and
 * a type but no address or port, so nothing can be dialled until the resolve
 * returns.
 */
class LanDiscoverer(private val context: Context) {

    interface Events {
        fun onFound(instanceId: String, host: String, port: Int, addressHash: Int)
        fun onLost(instanceId: String)
        fun onAvailabilityChanged(available: Boolean, reason: Unavailable?)
    }

    enum class Unavailable { PERMISSION_DENIED, NO_NETWORK }

    companion object {
        /** Must match the iOS side exactly or the two platforms never meet. */
        const val SERVICE_TYPE = "_kishorek-relay._tcp"

        /** TXT key carrying the truncated mesh address. */
        const val TXT_HASH = "h"

        /** TXT key carrying the per-run instance id. */
        const val TXT_INSTANCE = "i"
    }

    private val nsd: NsdManager? =
        context.getSystemService(Context.NSD_SERVICE) as? NsdManager

    private val connectivity: ConnectivityManager? =
        context.getSystemService(Context.CONNECTIVITY_SERVICE) as? ConnectivityManager

    var events: Events? = null

    private var registrationListener: NsdManager.RegistrationListener? = null
    private var discoveryListener: NsdManager.DiscoveryListener? = null
    private var networkCallback: ConnectivityManager.NetworkCallback? = null

    /** The name the system actually registered, which may be renamed on conflict. */
    private var registeredName: String? = null

    private val resolveQueue = ArrayDeque<NsdServiceInfo>()
    private var resolving = false

    // ---------------------------------------------------------- availability

    fun isAvailable(): Boolean = nsd != null && onWifi()

    fun unavailableReason(): Unavailable? = when {
        nsd == null -> Unavailable.PERMISSION_DENIED
        !onWifi() -> Unavailable.NO_NETWORK
        else -> null
    }

    /**
     * Wi-Fi or Ethernet only. Cellular is deliberately excluded: there are no
     * peers on a carrier network, and advertising there would be pure battery
     * cost for a discovery that can never succeed.
     */
    private fun onWifi(): Boolean {
        val manager = connectivity ?: return false
        val network = manager.activeNetwork ?: return false
        val capabilities = manager.getNetworkCapabilities(network) ?: return false
        return capabilities.hasTransport(NetworkCapabilities.TRANSPORT_WIFI) ||
            capabilities.hasTransport(NetworkCapabilities.TRANSPORT_ETHERNET)
    }

    fun watchNetwork() {
        if (networkCallback != null) return
        val manager = connectivity ?: return

        val callback = object : ConnectivityManager.NetworkCallback() {
            override fun onAvailable(network: Network) = publishAvailability()
            override fun onLost(network: Network) = publishAvailability()
            override fun onCapabilitiesChanged(
                network: Network,
                capabilities: NetworkCapabilities,
            ) = publishAvailability()
        }
        networkCallback = callback

        manager.registerNetworkCallback(
            NetworkRequest.Builder()
                .addTransportType(NetworkCapabilities.TRANSPORT_WIFI)
                .addTransportType(NetworkCapabilities.TRANSPORT_ETHERNET)
                .build(),
            callback,
        )
    }

    fun stopWatchingNetwork() {
        val callback = networkCallback ?: return
        networkCallback = null
        try {
            connectivity?.unregisterNetworkCallback(callback)
        } catch (_: IllegalArgumentException) {
            // Already unregistered.
        }
    }

    private fun publishAvailability() {
        events?.onAvailabilityChanged(isAvailable(), unavailableReason())
    }

    // ----------------------------------------------------------- advertising

    fun advertise(instanceId: String, addressHash: Int, port: Int) {
        val manager = nsd ?: return
        stopAdvertising()

        val info = NsdServiceInfo().apply {
            serviceName = instanceId
            serviceType = SERVICE_TYPE
            this.port = port
            setAttribute(TXT_INSTANCE, instanceId)
            // Unsigned hex rather than the raw integer: TXT values are bytes,
            // and a negative decimal would have to be parsed differently on
            // each platform.
            setAttribute(TXT_HASH, (addressHash.toLong() and 0xFFFFFFFFL).toString(16))
        }

        val listener = object : NsdManager.RegistrationListener {
            override fun onServiceRegistered(info: NsdServiceInfo) {
                registeredName = info.serviceName
            }

            override fun onRegistrationFailed(info: NsdServiceInfo, errorCode: Int) {
                events?.onAvailabilityChanged(false, Unavailable.NO_NETWORK)
            }

            override fun onServiceUnregistered(info: NsdServiceInfo) {
                registeredName = null
            }

            override fun onUnregistrationFailed(info: NsdServiceInfo, errorCode: Int) = Unit
        }

        registrationListener = listener
        try {
            manager.registerService(info, NsdManager.PROTOCOL_DNS_SD, listener)
        } catch (_: IllegalArgumentException) {
            registrationListener = null
        }
    }

    fun stopAdvertising() {
        val listener = registrationListener ?: return
        registrationListener = null
        registeredName = null
        try {
            nsd?.unregisterService(listener)
        } catch (_: IllegalArgumentException) {
            // Not registered. Nothing to undo.
        }
    }

    // -------------------------------------------------------------- browsing

    fun browse() {
        val manager = nsd ?: return
        if (discoveryListener != null) return

        val listener = object : NsdManager.DiscoveryListener {
            override fun onDiscoveryStarted(serviceType: String) = Unit

            override fun onStartDiscoveryFailed(serviceType: String, errorCode: Int) {
                discoveryListener = null
                events?.onAvailabilityChanged(false, Unavailable.NO_NETWORK)
            }

            override fun onStopDiscoveryFailed(serviceType: String, errorCode: Int) {
                discoveryListener = null
            }

            override fun onDiscoveryStopped(serviceType: String) {
                discoveryListener = null
            }

            override fun onServiceFound(info: NsdServiceInfo) {
                // Our own advertisement comes back to us; Dart filters on the
                // instance id too, but dropping it here saves a resolve on
                // every device in the room.
                if (info.serviceName == registeredName) return
                enqueueResolve(info)
            }

            override fun onServiceLost(info: NsdServiceInfo) {
                events?.onLost(info.serviceName)
            }
        }

        discoveryListener = listener
        try {
            manager.discoverServices(SERVICE_TYPE, NsdManager.PROTOCOL_DNS_SD, listener)
        } catch (_: IllegalArgumentException) {
            discoveryListener = null
        }
    }

    fun stopBrowsing() {
        val listener = discoveryListener ?: return
        discoveryListener = null
        resolveQueue.clear()
        try {
            nsd?.stopServiceDiscovery(listener)
        } catch (_: IllegalArgumentException) {
            // Discovery already stopped.
        }
    }

    // -------------------------------------------------------------- resolves

    @Synchronized
    private fun enqueueResolve(info: NsdServiceInfo) {
        resolveQueue.addLast(info)
        pumpResolves()
    }

    @Synchronized
    private fun finishResolve() {
        resolving = false
        pumpResolves()
    }

    private fun pumpResolves() {
        if (resolving) return
        val manager = nsd ?: return
        val next = resolveQueue.pollFirst() ?: return
        resolving = true

        @Suppress("DEPRECATION")
        manager.resolveService(
            next,
            object : NsdManager.ResolveListener {
                override fun onResolveFailed(info: NsdServiceInfo, errorCode: Int) {
                    // Routine. A device can vanish between being announced and
                    // being resolved, and mDNS caches outlive their devices.
                    finishResolve()
                }

                override fun onServiceResolved(info: NsdServiceInfo) {
                    publishResolved(info)
                    finishResolve()
                }
            },
        )
    }

    private fun publishResolved(info: NsdServiceInfo) {
        val host = info.host?.hostAddress ?: return
        val instanceId = attribute(info, TXT_INSTANCE) ?: info.serviceName ?: return
        if (instanceId == registeredName) return

        val hash = attribute(info, TXT_HASH)?.toLongOrNull(16)?.toInt() ?: return

        events?.onFound(instanceId, host, info.port, hash)
    }

    private fun attribute(info: NsdServiceInfo, key: String): String? =
        info.attributes[key]?.let { String(it, Charsets.UTF_8) }
}

File: app/ios/Runner/Wifi/LanDiscoverer.swift
import Foundation
import Network

/// Bonjour discovery on Apple platforms.
///
/// Two deliberate choices, both worth knowing before changing anything here.
///
/// **Bonjour, not a UDP beacon.** Since iOS 14, sending or receiving multicast
/// and broadcast requires the `com.apple.developer.networking.multicast`
/// entitlement, which Apple grants by application. Bonjour needs only a usage
/// string, and Android's NsdManager speaks the same protocol, so one mechanism
/// serves both platforms with no approval process.
///
/// **`NetService`, not `NWBrowser`.** The sockets belong to Dart, so this class
/// has to publish a service on a port it does not own and resolve a peer to a
/// plain IP and port. `NetService` does both directly. `NWBrowser` returns
/// endpoints that only resolve into an address by opening an `NWConnection`,
/// which would mean a second, pointless connection to every peer. `NetService`
/// is soft-deprecated as of iOS 15 but not removed; if it ever is, the
/// replacement is `NWBrowser` plus that extra connection.
final class LanDiscoverer: NSObject {

  /// Must match the Android side exactly or the two platforms never meet.
  static let serviceType = "_kishorek-relay._tcp"
  static let domain = "local."

  /// TXT keys, likewise shared with Android.
  static let txtHash = "h"
  static let txtInstance = "i"

  enum Unavailable {
    case permissionDenied
    case noNetwork
  }

  struct Found {
    let instanceId: String
    let host: String
    let port: Int
    let addressHash: UInt32
  }

  var onFound: ((Found) -> Void)?
  var onLost: ((String) -> Void)?
  var onAvailabilityChanged: ((Bool, Unavailable?) -> Void)?

  private var published: NetService?
  private var browser: NetServiceBrowser?

  /// Resolving services, held because `NetService` does no work unless someone
  /// keeps a strong reference to it for the duration of the resolve.
  private var resolving: [String: NetService] = [:]

  private var registeredName: String?

  private let monitor = NWPathMonitor()
  private let monitorQueue = DispatchQueue(label: "dev.kishorek.relay.wifi.path")

  private var onWifi = false

  /// True once a browse has failed in a way that looks like a refused
  /// local-network permission.
  ///
  /// iOS provides no API to ask. This is the best signal available, and it is
  /// a guess: a refused permission usually produces silence rather than an
  /// error, so the app can find itself unable to say why nothing is happening.
  private var browseFailed = false

  // ---------------------------------------------------------- availability

  func startMonitoring() {
    monitor.pathUpdateHandler = { [weak self] path in
      guard let self else { return }
      // Wi-Fi and wired only. Cellular carries no peers, and advertising there
      // would cost battery for a discovery that cannot succeed.
      let usable = path.status == .satisfied
        && (path.usesInterfaceType(.wifi) || path.usesInterfaceType(.wiredEthernet))
      guard usable != self.onWifi else { return }
      self.onWifi = usable
      self.onAvailabilityChanged?(self.isAvailable, self.unavailableReason)
    }
    monitor.start(queue: monitorQueue)
  }

  func stopMonitoring() {
    monitor.cancel()
  }

  var isAvailable: Bool { onWifi && !browseFailed }

  var unavailableReason: Unavailable? {
    if !onWifi { return .noNetwork }
    if browseFailed { return .permissionDenied }
    return nil
  }

  // ----------------------------------------------------------- advertising

  func advertise(instanceId: String, addressHash: UInt32, port: Int) {
    stopAdvertising()

    let service = NetService(
      domain: Self.domain,
      type: Self.serviceType,
      name: instanceId,
      port: Int32(port)
    )
    service.delegate = self

    var txt: [String: Data] = [:]
    txt[Self.txtInstance] = Data(instanceId.utf8)
    // Hex rather than a raw integer: TXT values are bytes, and a signed decimal
    // would have to be parsed differently on each platform.
    txt[Self.txtHash] = Data(String(addressHash, radix: 16).utf8)
    service.setTXTRecord(NetService.data(fromTXTRecord: txt))

    published = service
    registeredName = instanceId
    service.publish()
  }

  func stopAdvertising() {
    published?.stop()
    published = nil
    registeredName = nil
  }

  // -------------------------------------------------------------- browsing

  func browse() {
    guard browser == nil else { return }
    browseFailed = false

    let browser = NetServiceBrowser()
    browser.delegate = self
    self.browser = browser
    browser.searchForServices(ofType: Self.serviceType, inDomain: Self.domain)
  }

  func stopBrowsing() {
    browser?.stop()
    browser = nil
    resolving.removeAll()
  }
}

// MARK: - NetServiceDelegate

extension LanDiscoverer: NetServiceDelegate {

  func netServiceDidPublish(_ sender: NetService) {
    // The system renames on conflict, so the authoritative name is the one it
    // gives back. Filtering our own advertisement depends on having it right.
    registeredName = sender.name
  }

  func netService(_ sender: NetService, didNotPublish errorDict: [String: NSNumber]) {
    onAvailabilityChanged?(false, .noNetwork)
  }

  func netServiceDidResolveAddress(_ sender: NetService) {
    defer { resolving.removeValue(forKey: sender.name) }

    guard
      let data = sender.txtRecordData(),
      let host = Self.firstAddress(of: sender)
    else { return }

    let txt = NetService.dictionary(fromTXTRecord: data)
    let instanceId = txt[Self.txtInstance].flatMap { String(data: $0, encoding: .utf8) }
      ?? sender.name
    guard instanceId != registeredName else { return }

    guard
      let hashText = txt[Self.txtHash].flatMap({ String(data: $0, encoding: .utf8) }),
      let hash = UInt32(hashText, radix: 16)
    else { return }

    onFound?(
      Found(instanceId: instanceId, host: host, port: sender.port, addressHash: hash)
    )
  }

  func netService(_ sender: NetService, didNotResolve errorDict: [String: NSNumber]) {
    // Routine. mDNS caches outlive the devices behind them.
    resolving.removeValue(forKey: sender.name)
  }

  /// The first IPv4 address, as dotted quad.
  ///
  /// IPv4 only, deliberately. A link-local IPv6 address needs a scope id to be
  /// dialled, `dart:io` has no way to attach one, and every router that matters
  /// here hands out IPv4.
  private static func firstAddress(of service: NetService) -> String? {
    guard let addresses = service.addresses else { return nil }

    for data in addresses {
      let host: String? = data.withUnsafeBytes { raw -> String? in
        guard let base = raw.bindMemory(to: sockaddr.self).baseAddress else { return nil }
        guard base.pointee.sa_family == sa_family_t(AF_INET) else { return nil }

        var buffer = [CChar](repeating: 0, count: Int(NI_MAXHOST))
        guard
          getnameinfo(
            base,
            socklen_t(data.count),
            &buffer,
            socklen_t(buffer.count),
            nil,
            0,
            NI_NUMERICHOST
          ) == 0
        else { return nil }
        return String(cString: buffer)
      }
      if let host { return host }
    }
    return nil
  }
}

// MARK: - NetServiceBrowserDelegate

extension LanDiscoverer: NetServiceBrowserDelegate {

  func netServiceBrowser(
    _ browser: NetServiceBrowser,
    didFind service: NetService,
    moreComing: Bool
  ) {
    guard service.name != registeredName else { return }
    // Held strongly for the duration: a NetService with no owner stops
    // resolving and the delegate is never called.
    resolving[service.name] = service
    service.delegate = self
    service.resolve(withTimeout: 5)
  }

  func netServiceBrowser(
    _ browser: NetServiceBrowser,
    didRemove service: NetService,
    moreComing: Bool
  ) {
    resolving.removeValue(forKey: service.name)
    onLost?(service.name)
  }

  func netServiceBrowser(
    _ browser: NetServiceBrowser,
    didNotSearch errorDict: [String: NSNumber]
  ) {
    browseFailed = true
    self.browser = nil
    onAvailabilityChanged?(false, .permissionDenied)
  }
}
